### PR TITLE
First draft sign-in

### PR DESCRIPTION
**Pull request by zhoujc999**

### CODE REVIEW

**Review by Jaxonwht — APPROVED** [review]

File: client/package.json [hunk 1]
```diff
@@ -14,10 +14,11 @@
     "classnames": "^2.3.1",
     "react": "^18.2.0",
     "react-dom": "^18.2.0",
-    "react-scripts": "5.0.1",
+    "react-scripts": "4.0.3",
```
Comment (zhoujc999) [inline client/package.json:17]: changed to 4.0.3 to fix https://stackoverflow.com/questions/64557638/how-to-polyfill-node-core-modules-in-webpack-5

File: client/src/App.tsx [hunk 2]
```diff
@@ -13,13 +16,31 @@ const connectionStatusClassName = (connected: boolean) =>
 const App = () => {
   const [backendReady, setBackendReady] = useState(false);
   const [schedulerReady, setSchedulerReady] = useState(false);
+  const [token, setToken] = useState<null | string>(null);
```
Comment (zhoujc999) [inline client/src/App.tsx:19]: @Jaxonwht probably bad typing

File: client/src/Login.tsx [hunk 3]
```diff
@@ -0,0 +1,116 @@
+import React, { useState } from "react";
+import Web3 from "web3";
+
+let web3: Web3 | undefined = undefined; // Will hold the web3 instance
+
+const Login = ({ onLoggedIn }: any) => {
```
Comment (zhoujc999) [inline client/src/Login.tsx:6]: @Jaxonwht any type? maybe need to use interface or sth

File: client/src/Login.tsx [hunk 4]
```diff
@@ -0,0 +1,116 @@
...
+
+  const handleAuthenticate = ({
+    publicAddress,
+    signature,
+  }: {
+    publicAddress: string;
+    signature: string;
+  }) =>
+    Promise
+      .resolve // TODO: use axios to POST publicAddress, signature to backend.
```
Comment (zhoujc999) [inline client/src/Login.tsx:17]: @Jaxonwht TODO for backend

File: client/src/Login.tsx [hunk 5]
```diff
@@ -0,0 +1,116 @@
...
+
+      return { publicAddress, signature };
+    } catch (err) {
+      throw new Error("You need to sign the message to be able to log in.");
+    }
+  };
+
+  const handleSignup = (publicAddress: string) =>
+    Promise
+      .resolve // TODO: use axios to POST publicAddress to backend.
```
Comment (zhoujc999) [inline client/src/Login.tsx:44]: @Jaxonwht TODO for backend

File: client/src/Login.tsx [hunk 6]
```diff
@@ -0,0 +1,116 @@
...
+    const coinbase = await web3.eth.getCoinbase();
+    if (!coinbase) {
+      window.alert("Please activate MetaMask first.");
+      return;
+    }
+
+    const publicAddress = coinbase.toLowerCase();
+    setLoading(true);
+
+    // TODO: Look if user with current publicAddress is already present on backend
```
Comment (zhoujc999) [inline client/src/Login.tsx:81]: @Jaxonwht TODO for backend

File: client/package.json (hunk 1 above)
Comment (Jaxonwht) [reply client/package.json:17]: This is pretty bad.

File: client/src/App.tsx (hunk 2 above)
Comment (Jaxonwht) [reply client/src/App.tsx:19]: This is fine.

File: client/src/App.tsx [hunk 7]
```diff
@@ -13,13 +16,31 @@ const connectionStatusClassName = (connected: boolean) =>
...
 
+  useEffect(() => {
+    // Access token is stored in localstorage
+    const ls = window.localStorage.getItem(LS_KEY);
+    const token = ls && JSON.parse(ls);
+    setToken(token);
+  }, []);
+
+  const handleLoggedIn = (token: string) => {
+    localStorage.setItem(LS_KEY, JSON.stringify(token));
```
Comment (Jaxonwht) [inline client/src/App.tsx:35]: What even is `localStorage`. Uncompiled. `window.localStorage`.

File: client/src/Login.tsx (hunk 3 above)
Comment (Jaxonwht) [reply client/src/Login.tsx:6]: Create a `LoginProps`.

Comment (Jaxonwht) [reply client/src/Login.tsx:6]: interface LoginProps {
  onLoggedIn: () -> void;
}

File: client/src/Login.tsx (hunk 4 above)
Comment (Jaxonwht) [reply client/src/Login.tsx:17]: Okie

File: client/src/Login.tsx (hunk 5 above)
Comment (Jaxonwht) [reply client/src/Login.tsx:44]: Okie

File: client/src/Login.tsx [hunk 8]
```diff
@@ -0,0 +1,116 @@
...
+    }
+  };
+
+  const handleSignup = (publicAddress: string) =>
+    Promise
+      .resolve // TODO: use axios to POST publicAddress to backend.
+      // Backend add to user database and return address and nonce.
+      ()
+      .then((response) =>
+        JSON.parse('{"publicAddress":"' + publicAddress + '", "nonce":42}')
```
Comment (Jaxonwht) [inline client/src/Login.tsx:48]: nounce: 42?

File: client/src/Login.tsx [hunk 9]
```diff
@@ -0,0 +1,116 @@
...
+      );
+
+  const handleClick = async () => {
+    // Check if MetaMask is installed
+    if (!(window as any).ethereum) {
```
Comment (Jaxonwht) [inline client/src/Login.tsx:53]: window as unknown

File: client/src/Login.tsx (hunk 6 above)
Comment (Jaxonwht) [reply client/src/Login.tsx:81]: okie

File: client/src/Login.tsx [hunk 10]
```diff
@@ -0,0 +1,116 @@
...
+
+    // TODO: Look if user with current publicAddress is already present on backend
+    // Use axios GET on publicAddress. The backend return a single
+    // {publicAddress, nonce} object.
+    Promise.resolve()
+      .then((response) =>
+        JSON.parse('{"publicAddress":"' + publicAddress + '", "nonce":42}')
+      )
+      // If yes, retrieve it. If no, create it.
+      .then((users) => (users ? users : handleSignup(publicAddress)))
```
Comment (Jaxonwht) [inline client/src/Login.tsx:89]: Use await

**Review by Jaxonwht — APPROVED** [review]
LGTM!

**Pull request merged**